test: add basic pathfinding test cases

jdrueckert opened the pull request:
Extract many tests cases from Terasology/Behaviors#92 to tests on the pathfinding itself.

Most importantly, this allows us to run these tests as _unit tests_ instead of long-running integration tests using MTE.

jdrueckert committed:
test: add helper definitions for basic common test worlds and paths

jdrueckert committed:
test: add basic test cases for individual plugins and plugin combinations

jdrueckert committed:
test: add up and down again test for WalkLeapFall combination

jdrueckert committed:
fix: add intermediate steps to de-/ascending paths

jdrueckert committed:
test: add down and up again test case for WalkLeapFall combination

jdrueckert committed:
chore: disable failing tests to investigate them in follow-up

skaldarnar commented on src/test/java/org/terasology/flexiblepathfinding/FallingJPSTest.java:
```diff
@@ -0,0 +1,184 @@
+import org.terasology.flexiblepathfinding.plugins.basic.FallingPlugin;
+
+/**
+ * Falling movements are purely vertical (down) and require additional horizontal walking movements
```
The term "movement" or "movement type" is introduce one level higher in Behaviors for the actual physical movement input.

```suggestion
 * Falling describes purely vertical (down) paths and does not allow for horizontal steps.
 *
```

skaldarnar commented on src/test/java/org/terasology/flexiblepathfinding/FallingJPSTest.java:
```diff
@@ -0,0 +1,184 @@
+
+/**
+ * Falling movements are purely vertical (down) and require additional horizontal walking movements
+ * The following test cases only attempt to test that the falling plugin alone does not incorrectly find valid paths for basic movements
```
Can you add one positive test case to ensure that as well? Basically, finding a path from a position in the air one block down to a "grounded" block.

jdrueckert replied:
done

skaldarnar commented on src/test/java/org/terasology/flexiblepathfinding/LeapingJPSTest.java:
```diff
@@ -0,0 +1,185 @@
+import org.terasology.flexiblepathfinding.helpers.TestWorlds;
+import org.terasology.flexiblepathfinding.plugins.basic.LeapingPlugin;
+
+/**
+ * Leaping movements are purely vertical (up) and require additional horizontal walking movements
+ * The following test cases only attempt to test that the leaping plugin alone does not incorrectly find valid paths for basic movements
+ **/
```
Same clarification as above, plus the request to add one positive test case.

```suggestion
/**
 * Leaping describes purely vertical (up) paths and does not allow for horizontal steps.
 *
 * The following test cases (mostly) attempt to test that the leaping plugin alone does not incorrectly find valid paths that require horizontal steps.
 **/
```

skaldarnar commented on src/test/java/org/terasology/flexiblepathfinding/WalkingJPSTest.java:
```diff
@@ -0,0 +1,180 @@
+import org.terasology.flexiblepathfinding.helpers.TestWorlds;
+import org.terasology.flexiblepathfinding.plugins.basic.WalkingPlugin;
+
+public class WalkingJPSTest {
```
Add javadoc similar to to LeapingTests and FallingTests to explain that Walking should find paths with only horizontal steps, and that no vertical steps are allowed?

jdrueckert replied:
done

skaldarnar commented on src/test/java/org/terasology/flexiblepathfinding/helpers/TestPaths.java:
```diff
@@ -0,0 +1,152 @@
+package org.terasology.flexiblepathfinding.helpers;
+
+public final class TestPaths {
+    public static final String[] singleFlatStepNorthPath = {
```
Checkstyle wants you to write this as 
```suggestion
    public static final String[] SINGLE_FLAT_STEP_NORTH_PATH= {
```

not sure this is better... maybe we should disable checkstyle for the whole file?

jdrueckert replied:
I'm okay with both

jdrueckert committed:
doc: improve wording on leaping-only and falling-only test files

Co-authored-by: Tobias Nett <skaldarnar@googlemail.com>

jdrueckert committed:
chore: address review comments

- add documentation to WalkingJPSTest
- add positive test case to LeapingJPSTest and FallingJPSTest
- fix checkstyle issues

jdrueckert commented:
![](https://github.trello.services/images/mini-trello-icon.png) [FlexibleMovement as a replacement for PathFinding](https://trello.com/c/AGtBwude/189-flexiblemovement-as-a-replacement-for-pathfinding)
